### PR TITLE
feat: Modify `ZMLoginCodeRequestTranscoder` to support 2FA for login SQSERVICES-1341

### PR DESCRIPTION
**Pull request by agisilaos**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQSERVICES-1341" title="SQSERVICES-1341" target="_blank">SubtaskSQSERVICES-1341</a>  [iOS] SE: modify ZMLoginCodeRequestTranscoder
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

With this PR we modify the `ZMLoginCodeRequestTranscoder` to support Two Factor Authentication for Login. More specifically:

- Renamed `codeRequestSync` to `phoneVerificationCodeRequestSync`
- Added a new property `emailVerificationCodeRequestSync`
- in nextRequest we add an if check for the `current phase`
- In `requestForSingleRequestSync` method we do a check for which sync we do have and based on that we return the appropriate request.
- Added tests

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m [hunk 1]
```diff
@@ -30,8 +30,9 @@
 
 @interface ZMLoginCodeRequestTranscoder() <ZMSingleRequestTranscoder>
 
-@property (nonatomic, weak) ZMAuthenticationStatus *authenticationStatus;
-@property (nonatomic) ZMSingleRequestSync *codeRequestSync;
+@property (nonatomic, strong) ZMAuthenticationStatus *authenticationStatus;
```
Comment (KaterinaWire) [inline Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:33]: Why did you change it to `strong`?

File: Source/Synchronization/Transcoders/ZMLoginTranscoder.m [hunk 2]
```diff
@@ -150,6 +150,10 @@ - (ZMTransportRequest *)loginRequest
         payload[@"phone"] = credentials.phoneNumber;
         payload[@"code"] = credentials.phoneNumberVerificationCode;
     }
+    else if (credentials.email != nil && credentials.emailVerificationCode != nil) {
+        payload[@"email"] = credentials.email;
+        payload[@"verification-code"]  =credentials.emailVerificationCode;
```
Comment (KaterinaWire) [inline Source/Synchronization/Transcoders/ZMLoginTranscoder.m:155]:
```suggestion
        payload[@"verification-code"]  = credentials.emailVerificationCode;
```

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m [hunk 3]
```diff
@@ -63,11 +68,21 @@ - (ZMTransportRequest *)nextRequest
 
 - (ZMTransportRequest *)requestForSingleRequestSync:(__unused ZMSingleRequestSync *)sync;
 {
-    ZMTransportRequest *request = [[ZMTransportRequest alloc] initWithPath:@"/login/send"
+
+    if (sync == _emailVerificationCodeRequestSync) {
```
Comment (KaterinaWire) [inline Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:72]: Why not `sync == self.emailVerificationCodeRequestSync` ?

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m [hunk 4]
```diff
@@ -44,17 +45,21 @@ - (instancetype)initWithGroupQueue:(id<ZMSGroupQueue>)groupQueue authenticationS
     if (self != nil) {
         self.groupQueue = groupQueue;
         self.authenticationStatus = authenticationStatus;
-        self.codeRequestSync = [[ZMSingleRequestSync alloc] initWithSingleRequestTranscoder:self groupQueue:groupQueue];
-        [self.codeRequestSync readyForNextRequest];
+        self.phoneVerificationCodeRequestSync = [[ZMSingleRequestSync alloc] initWithSingleRequestTranscoder:self groupQueue:groupQueue];
+        self.emailVerificationCodeRequestSync = [[ZMSingleRequestSync alloc] initWithSingleRequestTranscoder:self groupQueue:groupQueue];
+        [self.phoneVerificationCodeRequestSync readyForNextRequest];
```
Comment (KaterinaWire) [inline Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:50]: Not sure do we need this?

File: Source/Synchronization/Transcoders/ZMLoginTranscoder.m [hunk 5]
```diff
@@ -150,6 +150,10 @@ - (ZMTransportRequest *)loginRequest
         payload[@"phone"] = credentials.phoneNumber;
         payload[@"code"] = credentials.phoneNumberVerificationCode;
     }
+    else if (credentials.email != nil && credentials.emailVerificationCode != nil) {
```
Comment (KaterinaWire) [inline Source/Synchronization/Transcoders/ZMLoginTranscoder.m:153]: I think we should add credentials.password here as well. And to the payload.

File: Tests/Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoderTests.m [hunk 6]
```diff
@@ -77,6 +77,17 @@ - (void)testThatItReturnsExpectedRequestWhenThereIsPhoneNumber
     XCTAssertEqualObjects(request, expectedRequest);
 }
 
+-(void)testThatItReturnsExpectedRequestWhenThereIsEmailVerificationCode
+{
+    NSString *email = @"test@wire.com";
+    [self.authenticationStatus prepareForRequestingEmailVerificationCodeForLogin:email];
+
+    ZMTransportRequest *expectedRequest = [[ZMTransportRequest alloc] initWithPath:@"/verification-code/send" method:ZMMethodPOST payload:@{@"email": email} authentication:ZMTransportRequestAuthNone];
```
Comment (KaterinaWire) [inline Tests/Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoderTests.m:85]: The test name says about the email verification code, which we don't have in the payload.

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m [hunk 7]
```diff
@@ -63,11 +68,21 @@ - (ZMTransportRequest *)nextRequest
...
 {
-    ZMTransportRequest *request = [[ZMTransportRequest alloc] initWithPath:@"/login/send"
+
+    if (sync == _emailVerificationCodeRequestSync) {
+        ZMTransportRequest *emailVerficationCodeRequest = [[ZMTransportRequest alloc] initWithPath:@"/verification-code/send"
+                                                                        method:ZMMethodPOST
+                                                                       payload:@{@"email": self.authenticationStatus.loginEmailThatNeedsAValidationCode}
+                                                                authentication:ZMTransportRequestAuthNone];
+        return emailVerficationCodeRequest;
+    } else {
```
Comment (johnxnguyen) [inline Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:78]: Probably we should have `else if (sync == self.phoneVerificationCodeRequestSync)`

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m (hunk 4 above)
Comment (agisilaos) [reply Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:50]: What do you mean? Are you referring to line 50? @KaterinaWire

File: Tests/Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoderTests.m (hunk 6 above)
Comment (agisilaos) [reply Tests/Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoderTests.m:85]: Maybe I need to find a better name for it. 🤔

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m (hunk 1 above)
Comment (agisilaos) [reply Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:33]: Because I would get errors like this
Screenshot 2022-03-14 at 09 55 50

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m (hunk 7 above)
Comment (agisilaos) [reply Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:78]: If we should have that `else if (sync == self.phoneVerificationCodeRequestSync)` what we would return as the default value? If we add such an else if we get an error like:

`Non-void function does not return a value in all control paths`

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m (hunk 4 above)
Comment (KaterinaWire) [reply Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:50]: Yes

File: Source/Synchronization/Transcoders/ZMLoginTranscoder.m [hunk 8]
```diff
@@ -150,6 +150,11 @@ - (ZMTransportRequest *)loginRequest
         payload[@"phone"] = credentials.phoneNumber;
         payload[@"code"] = credentials.phoneNumberVerificationCode;
     }
+    else if (credentials.email != nil && credentials.password != nil && credentials.emailVerificationCode != nil) {
```
Comment (KaterinaWire) [inline Source/Synchronization/Transcoders/ZMLoginTranscoder.m:153]: Since the first condition is `if (credentials.email != nil && credentials.password != nil) {`, we will never reach this condition. Try to check `credentials.emailVerificationCode != nil` first.

File: Tests/Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoderTests.m (hunk 6 above)
Comment (KaterinaWire) [reply Tests/Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoderTests.m:85]: Depends on what you want to test.

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m (hunk 7 above)
Comment (johnxnguyen) [reply Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:78]: Perhaps nil?

File: Tests/Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoderTests.m (hunk 6 above)
Comment (agisilaos) [reply Tests/Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoderTests.m:85]: I think I found one!

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m (hunk 4 above)
Comment (agisilaos) [reply Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:50]: Not sure either. @johnxnguyen thoughts?

**Comment by KaterinaWire** [comment]
@agisilaos @johnxnguyen  I think the person who added the comment should "resolve it" because it's hard to check it later. What do you think?

File: Source/Synchronization/Transcoders/ZMLoginTranscoder.m (hunk 8 above)
Comment (agisilaos) [reply Source/Synchronization/Transcoders/ZMLoginTranscoder.m]: @KaterinaWire I changed the order again. Please let me know if that was what you had in mind. I think that would be the right approach.

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m (hunk 4 above)
Comment (johnxnguyen) [reply Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:50]: It depends on what the method does. Can you check?

Comment (agisilaos) [reply Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:50]: @johnxnguyen The method is called 

```swift
- (instancetype)initWithGroupQueue:(id<ZMSGroupQueue>)groupQueue authenticationStatus:(ZMAuthenticationStatus *)authenticationStatus
{
    self = [super init];
     if (self != nil) {
         self.groupQueue = groupQueue;
         self.authenticationStatus = authenticationStatus;
         self.codeRequestSync = [[ZMSingleRequestSync alloc] initWithSingleRequestTranscoder:self groupQueue:groupQueue];
         [self.codeRequestSync readyForNextRequest];
         self.phoneVerificationCodeRequestSync = [[ZMSingleRequestSync alloc] initWithSingleRequestTranscoder:self groupQueue:groupQueue];
         self.emailVerificationCodeRequestSync = [[ZMSingleRequestSync alloc] initWithSingleRequestTranscoder:self groupQueue:groupQueue];
         [self.phoneVerificationCodeRequestSync readyForNextRequest];
     }
     return self;
 }
 ```

Comment (johnxnguyen) [reply Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:50]: Aren't we talking about `readyForNextRequest` though?

Comment (agisilaos) [reply Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:50]: Yeah... So regarding that:

```
/// Marks as need to request, even if it's already performing a request (will abort that request)
- (void)readyForNextRequest;
```

**Review by David-Henner — APPROVED** [review]

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m (hunk 4 above)
Comment (johnxnguyen) [reply Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:50]: It's still not clear to me why it's needed, but I suppose it's safer to keep it. If anything, I suggest we also add

```swift
[self.emailVerificationCodeRequestSync readyForNextRequest];
```

**Review by johnxnguyen — APPROVED** [review]

File: Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m (hunk 4 above)
Comment (agisilaos) [reply Source/Synchronization/Transcoders/ZMLoginCodeRequestTranscoder.m:50]: Done!
